csrc/ir/internal_nodes.cpp

```diff
@@ -3302,99 +3314,189 @@ _scaled_dot_product_flash_attention_meta(const at::Tensor& query) {
-    value = value.flatten(0, -4);
+  query = flattenBatchDims(query);
+  key = flattenBatchDims(key);
+  value = flattenBatchDims(value);
+  NVF_ERROR_EQ(query.dim(), 4);
+  NVF_ERROR_EQ(key.dim(), 4);
+  NVF_ERROR_EQ(value.dim(), 4);
+
+  at::Tensor attn_bias = bias;
+  if (mask.defined()) {
+    NVF_CHECK_EQ(mask.dtype(), at::kBool);
+    auto mask_bias =
+        at::where(mask, 0.0f, -std::numeric_limits<float>::infinity())
+            .to(query.dtype());
+    if (attn_bias.defined()) {
+      // Don't write `attn_bias += mask_bias` because the sum can be a larger
+      // shape than `attn_bias`.
+      attn_bias = attn_bias + mask_bias;
+    } else {
+      attn_bias = mask_bias;
+    }
+  }
+  if (attn_bias.defined()) {
+    attn_bias = flattenBatchDims(attn_bias);
   }
 
-  // ATen reference:
-  // https://github.com/pytorch/pytorch/blob/c27882ffa8c1c7e4cf8ebc6c2f879e5b6c8814ad/aten/src/ATen/native/transformers/attention.cpp#L680-L681
-  auto
-      [output,
-       log_sumexp,
-       cum_seq_q,
-       cum_seq_k,
-       query_seq_len,
-       key_seq_len,
-       philox_seed,
-       philox_offset,
-       debug_attn_mask] = query.is_meta()
-      ? spda_meta::_scaled_dot_product_flash_attention_meta(query)
-      : at::_scaled_dot_product_flash_attention(
-            query,
-            key,
-            value,
-            dropout_p,
-            is_causal,
-            /*return_debug_mask=*/false,
-            scale);
-
-  // If the inputs were padded, slice the output to restore the original
-  // size
-  if (output.size(-1) != last_dim_size) {
-    output = output.slice(-1, 0, last_dim_size);
-  }
-
-  auto unflatten_batch_dim = [](at::Tensor t,
-                                at::IntArrayRef batch_dims) -> at::Tensor {
-    at::DimVector new_shape(batch_dims);
-    auto non_batch_dims = t.sizes().slice(1);
-    new_shape.append(non_batch_dims.begin(), non_batch_dims.end());
-    return t.reshape(new_shape);
-  };
+  // 4D SDPA
+  auto [output, log_sumexp, philox_seed, philox_offset] = [&]() {
+    if (query.is_meta()) {
+      return sdpa_meta::_scaled_dot_product_attention_meta(query);
+    }
+
+    if (attn_bias.defined()) {
+      // This is a functional but suboptimal implementation for testing
+      // triangle attention:
+      // https://docs.nvidia.com/cuda/cuequivariance/api/generated/cuequivariance_torch.triangle_attention.html.
+      //
+      // To accommodate 5D, we combine `bias` and `mask` into `attn_bias`, a
+      // fully-allocated 5D tensor even though `bias` and `mask` have degenerate
+      // dimensions. Then, we flatten all inputs to 4D before running
+      // `at::_scaled_dot_product_attention_math`.
+      //
+      // This is suboptimal because:
+      // 1. Broadcasting `bias` and `mask` outside the kernel wastes GPU memory.
+      // 2. `at::_scaled_dot_product_attention_math` isn't fused. I
+      // didn't use `at::_scaled_dot_product_efficient_attention` because it
```

Comment:
For my information, what are the constraints required?

Reply:
I wish I remember all what I hit. For example, the sequence length has to be a multiple of 8. The sequence length in logsumexp is automatically padded to a multiple of 32, so the caller has to slice.